### DISCUSSION TITLE
textX matches link-reference even though it doesn't exist

### DISCUSSION THREAD

**Discussion by cosminpolifronie** [textX/textX#427] [General]
Hello! I am trying to migrate some grammars from Xtext to textX, mainly because of the abysmal documentation of the former.

The issue I'm hitting is that the link-reference is matching even though that reference doesn't exist.
In the case below, line `type UUID LogicalValues: 2 = abc; 5 = def;` will treat `LogicalValues` as a `Reference`, even though a `resource` entry with name `LogicalValues` doesn't exist in the model.

In the debug logs below look at lines:
`++ Match 'LogicalValues' at 62 => 'type UUID *LogicalValues*'`
`-- No match 'LogicalValues:' at 75 => 'icalValues*: 2 = abc; 5 =*'`
[debug.log](https://github.com/user-attachments/files/16337981/debug.log)

How should I go around fixing this?

Grammar:
```
`ServiceGenerator:
	services+=Service
;

Service:
	'service' name = CustomID '{'
		resources+=Resource
	'}'
;

Resource:
	'resource' name = CustomID 
	'{'
		properties+=Property
	'}'
;

Property:
	name = CustomID 
	type=ResourcePropertyType (reference=[Resource:CustomID])?
	('LogicalValues:' logicalValues+=LogicalValue)?
;

LogicalValue:
	index = INT '=' name = CustomID ';'?
;

ResourcePropertyType:
	'UUID' 		|
	'Reference'			
;

CustomID: 
	'n/a' | /[a-zA-Z_][0-9a-zA-Z_\-]*\b/
;
```

Model:
```
service test
{
     resource test1
     {
          type UUID LogicalValues: 2 = abc; 5 = def;
          value  UInt32   
     }

     resource test2
     {
          elementId  UUID
          functionOnDemandStatus  Enum:UInt32  LogicalValues: 0 = dummy; 1 = dummy2;
     }
}
```

**Comment by goto40** [comment]
Reference resolution happens after parsing. You need to disambiguate this, e.g. with a keyword like `->`: `type=ResourcePropertyType ('->' reference=[Resource:CustomID])? ('LogicalValues:' ...... `

**Reply by cosminpolifronie** [reply]
Does Xtext do this differently? Why did it work there?

**Comment by cosminpolifronie** [comment]
I took @goto40's advice and I made the case a bit more obvious. I changed the `Property` rule `type` entry to `type=ResourcePropertyTypeOrReference` and `ResourcePropertyTypeOrReference` looks like this:
```
ResourcePropertyTypeOrReference:
	ResourcePropertyType | 'Reference' reference=[Resource:CustomID]
;
```

I'm still wondering what Xtext does differently and why the previous way works there.

**Comment by goto40** [comment]
Xtext has a tokenizer which probably prevented your keyword to be used as ID

**Comment by goto40** [comment]
Textx has a packrat parser, which is context sensitive. No global lexer is present which makes many grammar aspects easier, in my option. However it differs from xtext or langium.

**Comment by igordejanovic** [comment]
As @goto40 already noted, PEG is different from CFGs. It is more an imperative style of grammar definition (alternative choice is ordered from left to right) while CFGs are more declarative. The problem you are facing is due to the fact that  textX uses on-the-fly (or context-aware) tokenization which is generally more powerful than doing tokenization upfront before parsing. Also, textX won't automatically prevent matching your keywords with liberal rules which accept string looking like keywords while other tools (like xText) might. This is also due to the power of context-aware lexing, i.e. textX won't try to match a keyword where it is not expected where upfront tokenization doesn't know the parsing context (unless the parser hack is used) so those tools use other techniques to prevent lexical ambiguities.

But, in your case *it is* expected to find a keyword `LogicalValues:` at the same place where `CustomID` could be matched as `CustomID` match is optional. Since you are matching this in a sequence and you are first trying `CustomID` which can match `LogicalValues` that will be consumed and the rest of the input, starting with `:` is not parsable anymore.

In addition to technique suggested by @goto40 another way to deal with this ambiguity is to make your `CustomID` match only if it is not a keyword, using negative lookahead. See section [Syntactic predicates](https://textx.github.io/textX/grammar.html?#syntactic-predicates) in the docs for an example how to do this. This technique doesn't require you to change the syntax of the language.

A simple fix without introducing new rules would be just:

```
Property:
	name = CustomID 
	type=ResourcePropertyType (!'LogicalValues:' reference=[Resource:CustomID])?
	('LogicalValues:' logicalValues+=LogicalValue)?
;
```
Notice the addition of `!'LogicalValues:'`

Hope this helps.
